backend/tracim_backend/lib/core/user.py

```diff
@@ -670,12 +681,66 @@ def authenticate(self, password: str, login: str, ldap_connector: "Connector" =
 
         raise AuthenticationFailed("Auth mechanism for this user is not activated")
 
+    def saml_authenticate(
+        self,
+        user: typing.Optional[User],
+        user_id: str,
+        name: str,
+        mail: str,
+    ) -> User:
+        """
+        Authenticate with ldap, return authenticated user or raise Exception
+        like WrongAuthTypeForUser, WrongLDAPCredentials, UserDoesNotExist
+        or UserAuthenticatedIsNotActive
+        :param user: user to check,, can be none if user not found, will try
+         to create new user if none but ldap auth succeed
+        :param user_id: user_id of the user
+        :param name: name extracted from saml attributes
+        :param mail: mail extracted from saml attributes
+        """
+        auth_type = AuthType.SAML
+
+        # TODO - M.L - 2023/09/06 - Change this to get saml response and isolate attributes
+
+        if not user:
+            try:
+                user = self.get_one_by_login(user_id)
+            except UserDoesNotExist:
+                _ = self.create_user(
+                    email=mail,
+                    username=user_id,
+                    name=name,
+                    auth_type=AuthType.SAML,
+                    do_save=True,
+                    do_notify=False,
+                )
+                transaction.commit()
+                user = self.get_one_by_login(user_id)
+        if user and user.auth_type not in [auth_type, AuthType.UNKNOWN]:
+            raise WrongAuthTypeForUser(
+                'User "{}" auth_type is {} not {}'.format(
+                    user_id, user.auth_type.value, auth_type.value
+                )
+            )
+
+        if user.is_deleted:
+            raise UserDoesNotExist("This user has been deleted")
+
+        if not user.is_active:
+            raise UserAuthenticatedIsNotActive("This user is not activated")
+
+        if user.auth_type == AuthType.UNKNOWN:
+            user.auth_type = auth_type
+        return user
```

Comment:
This function, like other similar functions, will attempt to find a user matching the provided user_id. If it can't find one, it will create a user with the provided information.

Reply:
@Millefeuille42 is there an external_id mechanism?

Reply:
Not currently